al_aws.js

```diff
@@ -34,6 +36,134 @@ var selfUpdate = function (callback) {
 };
 
 
+var selfConfigUpdate = function (callback) {
+    async.waterfall([
+        function(callback) {
+            getConfigChanges(function(err, config) {
+                callback(err, config)
+            });
+        },
+        function(newValues, callback) {
+            getCurrentConfig(function(err, currentConfig) {
+                callback(err, newValues, currentConfig)
+            });
+        },
+        function(newValues, currentConfig, callback) {
+            applyConfigChanges(newValues, currentConfig, function(err, newConfig) {
+                callback(err, newConfig, currentConfig)
+            })
+        },
+        function(newConfig, currentConfig, callback) {
+            if (isConfigDifferent(newConfig, currentConfig)) {
+                var lambda = new AWS.Lambda();
+                var updateConfig = filterDisallowedConfigParams(newConfig);
+                return lambda.updateFunctionConfiguration(updateConfig, callback);
+            } 
+            else {
+                callback(null);
+            }
+        }
+    ],
+    function(err, config) {
+        if (err) {
+            console.info('Lambda self-update config error: ', err);
+        } else {
+            if (config !== undefined) {
+                console.info('Lambda self-update config successful. Config: ', config);
+            }
+            else {
+                console.info('Lambda self-update config nothing to update');
+            }
+        }
+        callback(err, config);
+    });
+};
+
+function getConfigChanges(callback) {
+    var s3 = new AWS.S3();
+    var configName;
+    
+    // Use this default config in order to update old collectors
+    if (process.env.aws_lambda_update_config_name == undefined) {
+        configName = DEFAULT_CONFIG_NAME;
+    }
+    else {
+        configName = process.env.aws_lambda_update_config_name 
+    }
+    
+    var params = {
+        Bucket: process.env.aws_lambda_s3_bucket,
+        Key: configName
+    };
+    s3.getObject(params, function(err, object) {
+        if (err) {
+            return callback(err);
+        } else {
+            let config = JSON.parse(object.Body.toString());
+            return callback(null, config);
+        }
+    });
+}
+
+function getCurrentConfig(callback) {
+    var lambda = new AWS.Lambda();
+    var params = {
+        FunctionName: process.env.AWS_LAMBDA_FUNCTION_NAME
+    };
+    
+    lambda.getFunctionConfiguration(params, callback);
+}
+
+function applyConfigChanges(newValues, config, callback) {
+    var jsonConfig = JSON.stringify(config);
+    var newConfig = JSON.parse(jsonConfig);
+
+    try {
+        Object.keys(newValues).forEach(
+            function(item) {
+                let path = newValues[item]['path'];
+                let value = newValues[item]['value'];
+                changeObject(newConfig, path, value);
```

Comment:
what about using `jq` instead? it should help to simplify the code

Reply:
There are about 10 lines of code to change a value (`changeObject`). I don't really think it is hard to read it and we need an extra library for it.